delivery_auto_refresh/models/sale_order.py

```diff
@@ -1,21 +1,49 @@
 # Copyright 2018 Tecnativa - Pedro M. Baeza
+# Copyright 2021 Tecnativa - Carlos Roca
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 
-from odoo import api, models
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
 from odoo.tools import safe_eval
 
 
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
+    carrier_id = fields.Many2one(
+        compute="_compute_carrier_id", store=True, readonly=False
+    )
+    invoice_shipping_on_delivery = fields.Boolean(
+        string="Invoice Shipping on Delivery", copy=False
+    )
+
+    @api.depends("partner_id")
+    def _compute_carrier_id(self):
+        for rec in self:
+            if rec.partner_id:
+                if rec.partner_id.property_delivery_carrier_id:
+                    rec.carrier_id = rec.partner_id.property_delivery_carrier_id
+                else:
+                    carrier_ids = self.env["delivery.carrier"].search(
+                        [
+                            "|",
+                            ("company_id", "=", False),
+                            ("company_id", "=", self.company_id.id),
+                        ]
+                    )
+                    rec.carrier_id = min(carrier_ids, key=lambda c: c.sequence)
```

Comment:
The order is already on the model, so you don't need to reorder again. Just put `limit=1` in the search sentence. Anyway, if the order wouldn't be set, you can specify also in the search `order="field"` for auto-ordering on result (and also add `limit`).

Reply:
Test the onchange by the way.